src/dvc_objects/fs/path.py

```diff
@@ -135,3 +146,30 @@ def relparts(self, path, start=None):
 
     def as_posix(self, path):
         return path.replace(self.flavour.sep, posixpath.sep)
+
+    def split_version(self, path) -> Tuple[str, Optional[str]]:
+        if self.version_query_key:
+            parts = list(urlsplit(path))
+            query = parse_qs(parts[3])
+            if self.version_query_key in query:
+                version_id = first(query[self.version_query_key])
+                del query[self.version_query_key]
+                parts[3] = urlencode(query)
+            else:
+                version_id = None
+            return urlunsplit(parts), version_id
+        return path, None
+
+    def join_version(self, path: str, version_id: Optional[str]) -> str:
+        if not self.version_query_key:
+            raise ValueError("path does not support versioning")
+        parts = list(urlsplit(path))
+        query = parse_qs(parts[3])
+        if self.version_query_key in query:
+            raise ValueError("path already includes a version query")
+        parts[3] = f"versionid={version_id}" if version_id else ""
+        return urlunsplit(parts)
+
+    def version_path(self, path: str, version_id: Optional[str]) -> str:
+        path, _ = self.split_version(path)
+        return self.join_version(path, version_id)
```

Comment:
Can probably create `AzurePath` and add these methods there.

Reply:
This felt more like generic path methods to me. All of the filesystems that support versioning (s3, azure, gcs) use the querystring for specifying versions (and I'm not sure how else you would set a version in URL-equivalent paths?), so the only FS-specific part of versioning is the specific query param/key.

Reply:
and again, this is utils that I would expect to go in fsspec eventually, but for now it made more sense to me to have it in dvc-objects (shared across all of our fs's that support versioning) rather than fs specific path implementations

Reply:
How do you plan to use this, @pmrowla? Will it use `odb` or some APIs of `dvc-objects` and `dvc-data` that require version awareness?

If it's going to be a separate abstraction, I'd prefer it be implemented in individual plugins rather than here. 

> this is utils that I would expect to go in fsspec eventually

The `Path` is not upstreamed yet, so we should not add any more of our expectations/requirements to it.

Reply:
> How do you plan to use this, @pmrowla? Will it use `odb` or some APIs of `dvc-objects` and `dvc-data` that require version awareness?
> 
> If it's going to be a separate abstraction, I'd prefer it be implemented in individual plugins rather than here.

dvc-data is going to need to be aware of versioning metadata at some point. `version_id` is essentially just a per-file meta field, the same as `isexec`.

And when dealing with version-aware filesystems, a complete path includes the versioning query string.